clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/pubsub/input/PubSubSplitIterator.java

```diff
@@ -0,0 +1,256 @@
+    return pubSubConsumer.positionDifference(topicPartition, endPositionExclusive, currentPosition) > 1;
+  }
+
+  private void loadRecords() throws InterruptedException {
+    if (consumedRecordsIterator != null && consumedRecordsIterator.hasNext()) {
+      return;
+    }
+    Map<PubSubTopicPartition, List<DefaultPubSubMessage>> polled = Collections.emptyMap();
+    for (int i = 0; i < emptyRetryTimes; i++) {
+      polled = pubSubConsumer.poll(pollTimeoutMs);
+      if (!polled.isEmpty()) {
+        break;
+      }
+      Thread.sleep(emptySleepMs);
+    }
+    if (polled.isEmpty()) {
+      throw new VeniceException(
+          "Empty poll after " + emptyRetryTimes + " retries, tp=" + topicPartition + " pos=" + currentPosition + " end="
+              + endPositionExclusive + ". This may indicate no data available in the topic or partition.");
+    }
+    consumedRecordsIterator = Utils.iterateOnMapOfLists(polled);
+  }
+
+  public PubSubInputRecord next() throws IOException {
+    while (hasNext()) {
+      try {
+        loadRecords();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new VeniceException("Interrupted while fetching next message batch", e);
+      }
+
+      DefaultPubSubMessage pubSubMessage = consumedRecordsIterator.hasNext() ? consumedRecordsIterator.next() : null;
+
+      if (pubSubMessage == null) {
+        throw new IOException(
+            "Unable to read additional data. Partition " + topicPartition + " Current Position: " + currentPosition
+                + " End Position: " + endPositionExclusive);
+      }
+
+      currentPosition = pubSubMessage.getPosition();
+      long offset =
+          (mode == OffsetMode.LOGICAL_INDEX) ? (splitStartIndex + readSoFar) : currentPosition.getNumericOffset();
+      readSoFar += 1;
+
+      KafkaKey messageKey = pubSubMessage.getKey();
+      if (messageKey.isControlMessage()) {
+        // Skip control messages and record the skip.
+        controlMessagesSkipped += 1;
+        continue;
+      }
+
+      // Deliver a data record and record the delivery.
+      recordsDelivered += 1;
+      // Return a PubSubInputRecord that knows how to set itself
+      return new PubSubInputRecord(pubSubMessage, offset);
+    }
+    return null; // no more records
+  }
+
+  public float getProgress() {
+    if (targetCount <= 0) {
+      return 1.0f;
+    }
+    float progress = Math.min(1.0f, ((float) readSoFar / targetCount));
+
+    // Throttled progress logging (every >=1% change)
+    if (lastLoggedProgress < 0.0f || Math.abs(progress - lastLoggedProgress) >= 0.01f) {
+      LOGGER.info(
+          "PubSubSplitIterator progress for {}: read={} / {}, delivered={}, skipped={}, progress={}%",
+          topicPartition,
+          readSoFar,
+          targetCount,
+          recordsDelivered,
+          controlMessagesSkipped,
+          String.format("%.2f", progress * 100));
```

Comment:
if lastLoggedProgress was 99.5% , we may never log this line with progress = 100%.
not a big deal, but just wanted to highlight that.

Reply:
also, should we just print it in integer%, especially when logging happens at 1+% of progress? basically, do we want to see 1% 2% 4% 5%, or 1.1% 2.7% 4.3% 5.4%